### PR TITLE
Fix Upgrade Problematic and wrong Build State of SubJobs

### PR DESCRIPTION
**Pull request by ptrsny**
#135 
already includes #142

### CODE REVIEW

**Comment by vlsi** [comment]
>I will fix this in the future and hope, that noone will ever come in that situation!! ^^'

Just in case: I have such a case. Multijob performs integration testing (various configurations), and multiple multijob steps reuse the same "prepare db schema" job.

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 1]
```diff
@@ -170,25 +172,24 @@ private void addProjectFromBuilder(MultiJobProject project,
...
-
-										if (lastFailure != 0 && lastSuccess != 0) {
-											break;
-										}
-									}
-								}
+                            if (sb.getJobAlias().equals(projectConfig.getJobAlias()) &&
+                                    sb.getJobName().equals(projectConfig.getJobName())) {
+                                tmp_build = subProject.getBuildByNumber(sb.getBuildNumber());
+                                if (tmp_build != null) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: Technically speaking, I do not like this deeply nested if.
It would be much easier to follow in the following way:

```
if (tmp_build == null) {
  continue;
}
// other checks
```

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 2]
```diff
@@ -152,6 +152,8 @@ private void addProjectFromBuilder(MultiJobProject project,
             Item tli = Jenkins.getInstance().getItem(projectConfig.getJobName(), project.getParent(), AbstractProject.class);
             if (tli instanceof MultiJobProject) {
                 MultiJobProject subProject = (MultiJobProject) tli;
+                // TODO: only match those that have the same Alias name
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: It is not clear what this todo means. There is alias comparison below. Would you please rephrase?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 3]
```diff
@@ -170,25 +172,24 @@ private void addProjectFromBuilder(MultiJobProject project,
...
-											lastSuccess = tmp_build.getNumber();
-										}
-
-										if (lastFailure != 0 && lastSuccess != 0) {
-											break;
-										}
-									}
-								}
+                            if (sb.getJobAlias().equals(projectConfig.getJobAlias()) &&
+                                    sb.getJobName().equals(projectConfig.getJobName())) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: I would prefer 
```
if (not matches) {
  continue;
}
```

Otherwise this increases nesting level and makes it harder to follow

**Comment by ptrsny** [comment]
In the following Picture i illustrated what MultiJobPlugin is showing in its overview. I marked the information green that is correct, red that is always wrong and yellow the information that is correct in this configuration but can be wrong (like that marked red and yellow).

jenkins00


NOTICE: if there were another MultiJob under the Node MultiJob the information about it would be yellow/red also. the Node Multijob is just corrct, because it lies directly under the RootItem.

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 4]
```diff
@@ -145,63 +193,163 @@ private void addProjectFromBuilder(MultiJobProject project,
...
+                    addMultiProject(
+                            build.getProject(),
+                            (MultiJobProject)abstractProject,
+                            jobBuildState,
+                            phaseNestLevel + 1,
+                            currentPhaseName,
+                            out
+                    );
+                } else {
+                    BuildState jobBuildState = createBuildState(
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: This duplicates `if (subBuild == null)` branch.
That could you move `BuildState jobBuildState` above?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 5]
```diff
@@ -145,63 +193,163 @@ private void addProjectFromBuilder(MultiJobProject project,
...
+                            build,
+                            subBuild,
+                            jobBuildState,
+                            phaseNestLevel + 1,
+                            out
+                    );
+                }
+            } else {
+                AbstractBuild subBuild = null;
+                for ( SubBuild sb : build.getSubBuilds() ) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: This duplicates `abstractProject instanceof MultiJobProject` above. Could you move the common code out of identical `if` / `else` branches?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 6]
```diff
@@ -145,63 +193,163 @@ private void addProjectFromBuilder(MultiJobProject project,
...
+
+        for (PhaseJobsConfig projectConfig : subProjects) {
+            Item abstractProject = Jenkins.getInstance().getItem(
+                    projectConfig.getJobName(),
+                    build.getParent(),
+                    AbstractProject.class
+            );
+            if (abstractProject instanceof MultiJobProject) {
+                MultiJobBuild subBuild = null;
+                for (SubBuild sb : build.getSubBuilds()) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: Should `searchBuildnumberFromMultijobbuild` be used here?
`searchBuildnumberFromMultijobbuild` could return `SubBuild`.

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 7]
```diff
@@ -228,60 +389,71 @@ private BuildState createBuildState(BuildState parentBuildState,
...
-                                        .getPreviousBuildNumber());
-                    }
-                }
-            }
-        }
-        String displayName = project.getDisplayName();
-        if( alias != null )
+
+        String displayName = config.getJobName();
+        if( config.getJobAlias() != null )
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: Could you adjust spaces to `if (config.getJobAlias() != null)` ?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 8]
```diff
@@ -228,60 +389,71 @@ private BuildState createBuildState(BuildState parentBuildState,
...
 		{
-			if( !alias.equals("") ) {
-				displayName += " (" + alias + ")";
+			if( !config.getJobAlias().equals("") ) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: Adjust spaces please

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 9]
```diff
@@ -228,60 +389,71 @@ private BuildState createBuildState(BuildState parentBuildState,
...
+                        if (lastSuccessBuildNumber != 0) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
         if (previousParentBuild != null) {
             for (SubBuild subBuild : previousParentBuild.getSubBuilds()) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: `searchBuildnumberFromMultijobbuild` ?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 10]
```diff
@@ -228,60 +389,71 @@ private BuildState createBuildState(BuildState parentBuildState,
...
-                if (subBuild.getJobName().equals(project.getName())) {
+                if (subBuild.getJobName().equals(config.getJobName()) &&
+                        subBuild.getJobAlias().equals(config.getJobAlias())) {
                     previousBuildNumber = subBuild.getBuildNumber();
                 }
             }
         }
         if (lastParentBuild != null) {
             for (SubBuild subBuild : lastParentBuild.getSubBuilds()) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: `searchBuildnumberFromMultijobbuild` ?

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java (hunk 6 above)
Comment (ptrsny) [reply src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: searchBuildnumberFromMultijobbuild wants a specific result. this does not!

Comment (vlsi) [reply src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: The loops with equals/equals are extremely popular in this code and it is hard to follow which is which.

So either a couple of functions like `searchBuildnumberFromMultijobbuild` should be used or just `Result result` could be nullable so the method becomes universal.

Comment (ptrsny) [reply src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: a made the function universial. I will push my changes on monday next week.

File: src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java [hunk 11]
```diff
@@ -99,14 +100,23 @@ public FormValidation doCheckIncludeRegex(@QueryParameter String value)
 
     private void addTopLevelProject(MultiJobProject project,
             List<TopLevelItem> out) {
-        addMultiProject(null, project, createBuildState(project), 0, null, out);
+        if( project.getBuilds().isEmpty()) {
```
Comment (vlsi) [inline src/main/java/com/tikal/jenkins/plugins/multijob/views/MultiJobView.java]: space around opening brace is a bit odd

**Comment by cohencil** [comment]
Thanks a lot for this one!
